### PR TITLE
Improve clang-format support

### PR DESCRIPTION
**Pull request by BenjaminW3**
~This PR adds multiple things:~
* ~a `.clang-format` style that matches the current alpaka style as good as I was able to. We can easily change it later.~
* ~a CMake target `clang-format` for applying the `.clang-format` settings on the whole codebase (files in `include`, `test` and `example` folders with `hpp|h|cpp|c` extensions;  it does not format the `thirdParty` or other folders in the project.)~
* ~a CMake target `clang-format-check` which checks if the codebase is correctly formatted~
* ~a `CONTRIBUTING.md` file stating that clang-format is required for formatting the code. We can extend the file with other topics later.~

~Furthermore, the following things have been done:~
* ~the whole codebase was formatted~
* ~the CI will check that the code is correctly formatted, else the build will fail~

~On Windows the files to be formatted are not collected at configure-time because the number of files leads to command line length problems on Windows. Therefore the directories are iterated at runtime of the target. On linux files are collected at configure time.
CMake prints out a warning when the clang-format version is not 10.0 because this is the version enforced by CI.~

Improvements to the existing clang-format usage.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
Note: I will review it as soon as possible but I am currently ill and next week I have vocation.

**Comment by sbastrakov** [comment]
I love this change, but feel not qualified in this matter to review it.

File: script/clang-format-check.sh
Comment (psychocoderHPC) [inline script/clang-format-check.sh]: Can we add the command which the user must run to autoformat the code to fix the error.

File: script/clang-format-check.sh
Comment (psychocoderHPC) [inline script/clang-format-check.sh]: EOF missing

File: script/clang-format-check.bat
Comment (psychocoderHPC) [inline script/clang-format-check.bat]: EOF missing

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: script/clang-format-check.sh
Comment (BenjaminW3) [reply script/clang-format-check.sh]: This PR also adds a `clang-format` target which will do it for you.

**Comment by BenjaminW3** [comment]
Could someone try if this really works on linux?

File: script/clang-format-check.sh
Comment (psychocoderHPC) [reply script/clang-format-check.sh]: But you need clang 8 for that, right? If so we should mention it in the error message.

**Comment by psychocoderHPC** [comment]
I try to test it next week.

File: script/clang-format-check.sh
Comment (psychocoderHPC) [inline script/clang-format-check.sh]: please output `$comp_res` else it is not clear what is wrong

File: CMakeLists.txt
Comment (psychocoderHPC) [inline CMakeLists.txt]: you should search also for names like `clang-format-8` and maybe rate this `clang-format-8` higher than `clang-format`
I installed clang format 8 for my ubuntu 16.04 and do not have a binary `clang-format`

```
find_program(
    CLANG_FORMAT
    NAMES 
        "clang-format-9"
        "clang-format-8"
        "clang-format"
    PATHS ${CLANG_TOOLS_SEARCH_PATHS}
    DOC "Path to clang-format"
)
```

File: .clang-format
Comment (psychocoderHPC) [inline .clang-format]: Should be `AfterComma` else you will have first a comma and than the line break

File: .clang-format
Comment (psychocoderHPC) [inline .clang-format]: some shoudl be `AfterComma`

File: .clang-format
Comment (psychocoderHPC) [inline .clang-format]: maybe we need to add `PenaltyBreakTemplateDeclaration: 0` not sure if it is the default

File: script/clang-format-check.sh
Comment (psychocoderHPC) [inline script/clang-format-check.sh]: please make this file executable. `chmod 755` and than commit it

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]
note: clang format is also reordering the include order

File: .clang-format
Comment (psychocoderHPC) [inline .clang-format]: can we make this `FixNamespaceComments: true` I have very often hard times to see in which namespace a function is located if I scroll down.

File: .clang-format
Comment (psychocoderHPC) [inline .clang-format]: I suggest `IndentPPDirectives: AfterHash` or `BeforeHash` since we have a lot of preprocessor switches in the code it is sometimes hard to see where they start and end

```C++
//---------------------------------------HIP-----------------------------------
// __HIPCC__ is defined by hipcc (if either __HCC__ or __CUDACC__ is defined)
#if !defined(BOOST_LANG_HIP)
#    if defined(__HIPCC__) && (defined(__CUDACC__) || defined(__HCC__))
#        include <hip/hip_runtime.h>
//HIP defines "abort()" as "{asm("trap;");}", which breaks some kernels
#        undef abort
// there is no HIP_VERSION macro
#        define BOOST_LANG_HIP BOOST_VERSION_NUMBER_AVAILABLE
#        if defined(BOOST_LANG_CUDA) && BOOST_LANG_CUDA
#            undef BOOST_LANG_CUDA
#            define BOOST_LANG_CUDA BOOST_VERSION_NUMBER_NOT_AVAILABLE
#        endif
#    else
#        define BOOST_LANG_HIP BOOST_VERSION_NUMBER_NOT_AVAILABLE
#    endif
#endif
```

**Comment by psychocoderHPC** [comment]
`[&](TArgs const & ... args)` is changed by clang format into `[&](TArgs const &... args)` but can IMO not be avoided.

File: script/clang-format-check.sh
Comment (BenjaminW3) [reply script/clang-format-check.sh]: I will extend the error message.

File: script/clang-format-check.sh
Comment (BenjaminW3) [reply script/clang-format-check.sh]: What do you get from looking at the return code when you already know that it was nonzero?

File: CMakeLists.txt
Comment (BenjaminW3) [reply CMakeLists.txt]: Will add `clang-format-8`

File: .clang-format
Comment (BenjaminW3) [reply .clang-format]: I saw some really strange behaviour when indenting the preprocessor directives. Maybe it works now. However, this PR is not about the specific `.clang-format` file but about the tools. We can always enable this later.

**Comment by BenjaminW3** [comment]
@psychocoderHPC Did the `make clang-format` target work for you?

File: script/clang-format-check.sh
Comment (psychocoderHPC) [reply script/clang-format-check.sh]: This variable contains the diff and show you all the wrong formated lines.

**Comment by psychocoderHPC** [comment]
Yes make clang-format works if I change the find method like like I wrote above

File: script/clang-format-check.sh
Comment (BenjaminW3) [reply script/clang-format-check.sh]: I added printing the variable content

File: .clang-format
Comment (psychocoderHPC) [reply .clang-format]: sry I added the wrong keyword here: It must be `AfterColon` the suggested `AfterComma` is not existing.

File: .clang-format
Comment (psychocoderHPC) [reply .clang-format]: sry I added the wrong keyword here: It must be `AfterColon` the suggested `AfterComma` is not existing.

File: .clang-format
Comment (psychocoderHPC) [inline .clang-format]: This and the next line must be `AfterColon` I added the wrong keyword in my previous suggestion.

**Review by psychocoderHPC — DISMISSED** [review]

**Comment by ax3l** [comment]
You might want to add to your `CONTRIBUTING.md` how to add this check as a (client side) [git pre-commit hook](https://git-scm.com/book/pl/v2/Customizing-Git-Git-Hooks).

Alternatively, you can also rely on projects such as https://github.com/pre-commit/pre-commit with a little `.yaml` file and a lightweight python dependency.

Just an idea to make this more comfortable.

**Comment by psychocoderHPC** [comment]
@BenjaminW3  How would we go on with this PR. as @ax3l mentioned we need at least information in `CONTRIBUTION.md`  about the tools and requirements.

**Comment by BenjaminW3** [comment]
There is not yet a requirement to use formatting nor is any code fotmatted yet. I would only add documentation when it is finally enabled which is not part of this PR.

**Comment by psychocoderHPC** [comment]
I added WIP to the PR name

**Comment by BenjaminW3** [comment]
My original intention was to get this merged as is without adding formatting requirements. I will think about if we should format the source code together with the tools and the CI enforcement getting merged.

**Comment by BenjaminW3** [comment]
I updated the PR and the descritption on the top.

**Comment by BenjaminW3** [comment]
Ready to be approved/merged!

**Comment by ax3l** [comment]
@BenjaminW3 little conflict

**Comment by sbastrakov** [comment]
@BenjaminW3 @psychocoderHPC what is the status of this PR?

**Comment by BenjaminW3** [comment]
It is still my goal to integrate this. I will try to update it when I find some time.

**Comment by BenjaminW3** [comment]
I split off a refactoring I did in this PR and created #976.

**Comment by BenjaminW3** [comment]
This PR originally wanted to add a test that checks whether all files are correctly formatted.
Now that we have support for github actions I am not sure if we should still enforce this for all PRs.
As a first step I will definitely add the `.clang-format` file and format the code base once.
After doing so, we have the following options:
1) trust in developers to format the code before merging and do not care if some lines are not correctly formatted
2) Add a check into PR build which ensures that the code is correctly formatted before it can be merged
3) Use an action that automatically runs a script after merging which formats all the files and adds a commit on top if there are any unformatted files

We could probably extend option 2 (but I do not yet have any experience doing so):
2b) Extend 2) by adding a bot command to PRs like `!format!` which you can write as a PR comment. This starts an action that automatically adds a commit with correctly formatted code to the PR.

@alpaka-group/alpaka-maintainers What do you think?
I would like to start with 1) so that we get a baseline.

**Comment by SimeonEhrig** [comment]
I would suggest option two. The reason is really simple. It's quite likely that you get fake diffs using clang-format. If a developer runs clang-format over an entire file and parts of it are not formatted correctly before, diffs are created on code lines that the developer has never touched.

**Comment by psychocoderHPC** [comment]
> * trust in developers to format the code before merging and do not care if some lines are not correctly formatted

I do not prefer this option. Same reason @SimeonEhrig said (code changes in not touched code parts). I like the option two more, were we check the format in a CI job and provide the user with the command he needs execute to format all code.


> We could probably extend option 2 (but I do not yet have any experience doing so):
2b) Extend 2) by adding a bot command to PRs like !format! which you can write as a PR comment. This starts an action that automatically adds a commit with correctly formatted code to the PR.

A bot would be the second step of the CI job where the code is only checked. This would allow us as maintainer to fix format issues  without user interaction.

**Comment by bernhardmgruber** [comment]
I would like to revive this PR and at least get the .clang-format file merged. Are there still any objections against the state of the .clang-format checks?

Should we create a separate PR and just add the .clang-format file for now?

**Comment by BenjaminW3** [comment]
I would also like to get the two `.clang-format` files merged and the repository formatted once even without checks in CI.
A new PR should be good for this

**Comment by psychocoderHPC** [comment]
> I would also like to get the two `.clang-format` files merged and the repository formatted once even without checks in CI.
> A new PR should be good for this

We should not forget to say which clang-format version we require. There are small differences.
I would like to see a helper script we can run to be sure all folders with cpp and h files will always be formated.

**Comment by bernhardmgruber** [comment]
> We should not forget to say which clang-format version we require. There are small differences.
> I would like to see a helper script we can run to be sure all folders with cpp and h files will always be formated.

While there are indeed small differences, they should be negligible. It should be rare that we run clang-format on the entire codebase. I recently found out that you can run `git clang-format ...` to just format your current changes, your changes on a branch etc. That keeps formatting local. So essentially, a user can run `git clang-format develop` on their branch and have all their changes, since they branched off develop, formatted.

For formatting everything, there is already a guide here: https://github.com/ComputationalRadiationPhysics/mallocMC/blob/dev/CONTRIBUTING.md

**Comment by psychocoderHPC** [comment]
@BenjaminW3 What is missing to finish this PR? I would vote for bringing clang format in because we have more and more developers who working on various projects and formatting all correct is therefore impossible.

**Comment by bernhardmgruber** [comment]
If finishing this PR is too hard, I would be super happy, if we could just add a new PR with just the `.clang-format` file. The CI checks can be postponed for later IMO.

**Comment by psychocoderHPC** [comment]
Info: the final clang format file we will be using for all our projects will be voted https://github.com/ComputationalRadiationPhysics/contributing/pull/39

**Comment by BenjaminW3** [comment]
I removed the parts of this PR which now have been added by other PRs. There are still some improvements left.
Please approve ;-)

**Review by bernhardmgruber — APPROVED** [review]

**Pull request merged**
